File: src/daegun/week29/외벽 점검.swift
/*
 시간 초과
 */

import Foundation

func solution(_ n:Int, _ weak:[Int], _ dist:[Int]) -> Int {
    
    var allFriends: [[Int]] = []
    let visited: [Bool] = [Bool](repeating: false, count: dist.count)

    // 친구 순서 모든 경우의 수
    func permutation(depth: Int, visited: [Bool], arr: [Int]) {
        if depth == dist.count {
            allFriends.append(arr)
        }
        var visited = visited
        for i in 0..<dist.count {
            if visited[i] {
                continue
            }
            
            visited[i] = true
            permutation(depth: depth+1, visited: visited, arr: arr+[dist[i]])
            visited[i] = false
        }
    }
    permutation(depth: 0, visited: visited, arr: [])
    
    // 배치 시작 위치 모든 경우의 수
    var orders = [weak]
    for _ in 0..<weak.count-1 {
        var last = orders.last!
        let first = last.removeFirst()
        last.append(first + n)
        orders.append(last)
    }
    
    var result = dist.count + 1
    func check(_ order: [Int], _ friends: [Int]) {
        
        var count = 0
        var start = 0
        
        for end in 1..<order.count {
            
            guard start < end else { continue }
            
            if order[end] - order[start] < friends[count] {
                continue
                
            } else if order[end] - order[start] == friends[count] {
                continue
                
            } else {
                count += 1
                start = end + 1
            }
        }
        
        if start < order.count { count += 1 }
        
        result = min(result, count)
    }
    
    for order in orders {
        for friends in allFriends {
            
            check(order, friends)
        }
    }
    
    return (dist.count + 1 > result ? result : -1)
}

File: src/taeeun/week29/외벽 점검.cpp
#include <string>
#include <vector>
#include <algorithm>


using namespace std;

int solution(int n, vector<int> weak, vector<int> dist) {
    int answer = 101;
    //1. 원형을 일직선으로 바꿔줌
    int m=weak.size();
    weak.resize(m*2);
    for (int i=m; i<2*m; i++) {
        weak[i]=weak[i-m]+n;
    }
    
    sort(dist.begin(), dist.end());
    
    do {
        for (int i=0; i<m; i++) {
            int start=weak[i];
            int end=weak[i+m-1];
            for (int j=0; j<dist.size(); j++) {
                start+=dist[j];
                if (start>=end) {
                    answer=min(answer,j+1);
                    break;
                }
                
                //현재 지점에서 다음 가까운 취약 지점으로 이동
                for (int k=i+1; k<m; k++) {
                    if (weak[k]>start) {
                        start=weak[k];
                        break;
                    }
                }
            }        
        }
        
    }while(next_permutation(dist.begin(), dist.end()));
    
    if (answer==101) return -1;
    
    return answer;
}